### ISSUE TITLE
Cannot join secondary to single node formation

### ISSUE THREAD

**Issue by Bouke** [citusdata/pg_auto_failover#262]
As upgrading my formation has failed (see other issue), I'm rebuilding my formation. So far that's not going all to well. I've removed the formation and all nodes belonging to the formation. Then I'm rejoining the primary:

```
bash-4.4$ pg_autoctl create postgres --pgdata /var/lib/pgsql/12/data --nodename `hostname --fqdn` --monitor postgres://autoctl_node@tools.cb.local/pg_auto_failover --formation acceptance --auth trust --no-ssl
20:06:23 INFO  Found pg_ctl for PostgreSQL 12.2 at /usr/pgsql-12/bin/pg_ctl
20:06:23 ERROR Failed to open file "/var/lib/pgsql/12/data/postmaster.pid": No such file or directory
20:06:23 ERROR Failed to open file "/var/lib/pgsql/12/data/postmaster.pid": No such file or directory
20:06:23 WARN  A postgres directory already exists at "/var/lib/pgsql/12/data", registering as a single node
20:06:23 INFO  Registered node acc-sql01.cb.local:5432 with id 15 in formation "acceptance", group 0, state "single"
20:06:23 INFO  Writing keeper state file at "/var/lib/pgsql/.local/share/pg_autoctl/var/lib/pgsql/12/data/pg_autoctl.state"
20:06:23 ERROR Failed to open file "/var/lib/pgsql/12/data/postmaster.pid": No such file or directory
20:06:23 ERROR Failed to open file "/var/lib/pgsql/12/data/postmaster.pid": No such file or directory
20:06:23 ERROR Failed to open file "/var/lib/pgsql/12/data/postmaster.pid": No such file or directory
20:06:23 INFO  Writing keeper init state file at "/var/lib/pgsql/.local/share/pg_autoctl/var/lib/pgsql/12/data/pg_autoctl.init"
20:06:23 INFO  Successfully registered as "single" to the monitor.
20:06:23 INFO  FSM transition from "init" to "single": Start as a single node
20:06:23 INFO  Initialising postgres as a primary
20:06:23 ERROR Failed to open file "/var/lib/pgsql/12/data/postmaster.pid": No such file or directory
20:06:23 INFO  Postgres is not running, starting postgres
20:06:23 INFO   /usr/pgsql-12/bin/pg_ctl --pgdata /var/lib/pgsql/12/data --options "-p 5432" --options "-h *" --wait start
20:06:23 INFO  CREATE DATABASE postgres;
20:06:23 INFO  The database "postgres" already exists, skipping.
20:06:23 INFO  The user "pgautofailover_monitor" already exists, skipping.
20:06:23 INFO  The user "pgautofailover_replicator" already exists, skipping.
20:06:23 INFO  Transition complete: current state is now "single"
20:06:23 INFO  Keeper has been succesfully initialized.
```

So far, so good. The monitor also lists the node as single node member:
```
pg_autoctl show state --formation=acceptance                                                                                                                                                             Tue May 12 20:10:10 2020

              Name |   Port | Group |  Node |     Current State |    Assigned State
-------------------+--------+-------+-------+-------------------+------------------
acc-sql01.cb.local |   5432 |     0 |    15 |            single |            single
```

Next up, joining the secondary:
```
pg_autoctl create postgres --pgdata /var/lib/pgsql/12/data --nodename `hostname --fqdn` --monitor postgres://autoctl_node@tools.cb.local/pg_auto_failover --formation acceptance --auth trust --no-ssl
20:07:44 ERROR Failed to open file "/var/lib/pgsql/12/data/postmaster.pid": No such file or directory
20:07:44 INFO  Continuing from a previous `pg_autoctl create` failed attempt
20:07:44 INFO  PostgreSQL state at registration time was: PGDATA does not exists
20:07:44 INFO  FSM transition from "wait_standby" to "catchingup": The primary is now ready to accept a standby
20:07:44 INFO  The primary node returned by the monitor is acc-sql01.cb.local:5432
20:07:44 INFO  Initialising PostgreSQL as a hot standby
20:07:44 INFO  Running /usr/pgsql-12/bin/pg_basebackup -w -d 'application_name=pgautofailover_standby_11 host=acc-sql01.cb.local port=5432 user=pgautofailover_replicator' --pgdata /var/lib/pgsql/12/backup/acc-sql02.cb.local -U pgautofailover_replicator --write-recovery-conf --max-rate 100M --wal-method=stream --slot pgautofailover_standby ...
20:07:47 INFO  pg_basebackup: initiating base backup, waiting for checkpoint to complete
20:07:47 INFO  pg_basebackup: checkpoint completed
20:07:47 INFO  pg_basebackup: write-ahead log start point: 4/BB000028 on timeline 175
20:07:47 INFO  pg_basebackup: starting background WAL receiver
20:07:47 INFO       0/208696 kB (0%), 0/1 tablespace (.../acc-sql02.cb.local/backup_label)
20:07:47 INFO  pg_basebackup: error: could not send replication command "START_REPLICATION": ERROR:  replication slot "pgautofailover_standby" does not exist
20:07:47 INFO   49822/208696 kB (23%), 0/1 tablespace (...-sql02.cb.local/base/41119/49311)
20:07:47 INFO  154416/208696 kB (73%), 0/1 tablespace (...-sql02.cb.local/base/41119/49317)
20:07:47 INFO  208707/208707 kB (100%), 0/1 tablespace (...sql02.cb.local/global/pg_control)
20:07:47 INFO  208707/208707 kB (100%), 1/1 tablespace
20:07:47 INFO  pg_basebackup: write-ahead log end point: 4/BB000B08
20:07:47 INFO  pg_basebackup: waiting for background process to finish streaming ...
20:07:47 INFO  pg_basebackup: error: child process exited with exit code 1
20:07:47 INFO  pg_basebackup: removing contents of data directory "/var/lib/pgsql/12/backup/acc-sql02.cb.local"
20:07:47 ERROR Failed to run pg_basebackup: exit code 1
20:07:47 ERROR Failed initialise standby server, see above for details
20:07:47 ERROR Failed to transition from state "wait_standby" to state "catchingup", see above.
```

The following error seems relevant:

> ERROR:  replication slot "pgautofailover_standby" does not exist

And yes, on the primary there's no such slot:
```
psql
psql (12.2)
Type "help" for help.

postgres=# SELECT slot_name, slot_type, active FROM pg_replication_slots;
 slot_name | slot_type | active
-----------+-----------+--------
(0 rows)
```

Am I missing something in the setup?

**Comment by Bouke** [comment]
I finally managed to join the secondary by removing the the local state:
```
rm -rf /var/lib/pgsql/.local/share/pg_autoctl
```

I'm confused.

**Comment by DimCitus** [comment]
First, as you know, when you create the secondary the pg_autoctl process on the primary must be running. I believe that it was the case and you didn't report about it, but wanted to stress that point. The replication slot is created on the primary by the primary node's pg_autoctl process when transitioning from state single to state wait_primary.

When dropping a secondary node you can use `pg_autoctl drop node ---destroy` if you never want to re-use the existing configuration and local data directory. Otherwise it's more like “detaching” a node and keeping the local Postgres data directory for other purposes. In other words, by default we won't delete your data.

Now, we fixed some replication slots bugs in release 1.3, but I would not expect that bug anyway in 1.2, if I remember correctly. Also fixing the secondary by removing its internal on-disk state is not the expected procedure. Has the file been re-created?

**Comment by DimCitus** [comment]
I believe this issue has been fixed in the recent versions of pg_auto_failover (1.4.2), so closing this issue now. Feel free to open again with more information if needed.
